### PR TITLE
Add object id/address/transaction digest validity test #SUI-55

### PR DESCRIPTION
**Pull request by xydas97**
Work for implementation of ticket https://mysten.atlassian.net/browse/SUI-55?atlOrigin=eyJpIjoiMTczNDQwMjdhOGNjNDFhZThhZmZmYmU5NDYxODYwNGUiLCJwIjoiaiJ9

### CODE REVIEW

**Comment by xydas97** [comment]
ts-sdk: check if object id /address/transaction digest is valid before querying fullnode

File: sdk/typescript/src/providers/json-rpc-provider.ts [hunk 1]
```diff
@@ -312,8 +316,11 @@ export class JsonRpcProvider extends Provider {
   }
 
   // Objects
-  async getObjectsOwnedByAddress(address: string): Promise<SuiObjectInfo[]> {
+  async getObjectsOwnedByAddress(address: SuiAddress): Promise<SuiObjectInfo[]> {
     try {
+      if (!isValidSuiAddress(normalizeSuiAddress(address)) || !address) {
```
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:321]:
```suggestion
      if (!address || !isValidSuiAddress(normalizeSuiAddress(address))) {
```
nit: can we put `!address` at the front`?

File: sdk/typescript/src/providers/json-rpc-provider.ts [hunk 2]
```diff
@@ -376,8 +383,11 @@ export class JsonRpcProvider extends Provider {
     )) as GetObjectDataResponse[];
   }
 
-  async getObjectsOwnedByObject(objectId: string): Promise<SuiObjectInfo[]> {
+  async getObjectsOwnedByObject(objectId: ObjectId): Promise<SuiObjectInfo[]> {
     try {
+      if (!isValidSuiObjectId(normalizeSuiObjectId(objectId)) || !objectId) {
```
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:388]: ditto

File: sdk/typescript/src/providers/json-rpc-provider.ts [hunk 3]
```diff
@@ -391,8 +401,11 @@ export class JsonRpcProvider extends Provider {
     }
   }
 
-  async getObject(objectId: string): Promise<GetObjectDataResponse> {
+  async getObject(objectId: ObjectId): Promise<GetObjectDataResponse> {
     try {
+      if (!isValidSuiObjectId(normalizeSuiObjectId(objectId)) || !objectId) {
```
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:406]: ditto

File: sdk/typescript/src/providers/json-rpc-provider.ts [hunk 4]
```diff
@@ -404,12 +417,12 @@ export class JsonRpcProvider extends Provider {
...
 
-  async getObjectRef(objectId: string): Promise<SuiObjectRef | undefined> {
+  async getObjectRef(objectId: ObjectId): Promise<SuiObjectRef | undefined> {
     const resp = await this.getObject(objectId);
     return getObjectReference(resp);
   }
 
-  async getObjectBatch(objectIds: string[]): Promise<GetObjectDataResponse[]> {
+  async getObjectBatch(objectIds: ObjectId[]): Promise<GetObjectDataResponse[]> {
     const requests = objectIds.map((id) => ({
```
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:426]: We should also validate each of the `objectIds` here. If one of them is invalid, feel free to throw an error that tells which of the `objectIds` is invalid.

File: sdk/typescript/src/providers/json-rpc-provider.ts [hunk 5]
```diff
@@ -462,6 +475,9 @@ export class JsonRpcProvider extends Provider {
     ];
 
     try {
+      if (!isValidSuiObjectId(normalizeSuiObjectId(objectID)) || !objectID) {
```
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:478]: ditto

File: sdk/typescript/src/providers/json-rpc-provider.ts [hunk 6]
```diff
@@ -490,6 +506,9 @@ export class JsonRpcProvider extends Provider {
       },
     ];
     try {
+      if (!isValidSuiAddress(normalizeSuiAddress(addressID)) || !addressID) {
```
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:509]: ditto

File: sdk/typescript/src/providers/json-rpc-provider.ts [hunk 7]
```diff
@@ -507,6 +526,9 @@ export class JsonRpcProvider extends Provider {
     digest: TransactionDigest
   ): Promise<SuiTransactionResponse> {
     try {
+      if (!isValidTransactionDigest(digest)) {
```
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:529]: We also need to do the same validation for `getTransactionWithEffectsBatch` on line 546

**Review by 666lcz — CHANGES_REQUESTED** [review]
Thanks for working on this! Overall looks good with a few places to fix.

File: sdk/typescript/src/providers/json-rpc-provider.ts [hunk 8]
```diff
@@ -523,24 +550,28 @@ export class JsonRpcProvider extends Provider {
...
   async getTransactionWithEffectsBatch(
     digests: TransactionDigest[]
   ): Promise<SuiTransactionResponse[]> {
-    const requests = digests.map((d) => ({
-      method: 'sui_getTransaction',
-      args: [d],
-    }));
     try {
+      const requests = digests.map((d) => {
+        if (!isValidTransactionDigest(d, "base64") && !isValidTransactionDigest(d, "base58")) {
```
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:559]:
```suggestion
        if (!isValidTransactionDigest(d, "base58")) {
```
We have released `0.18.0` on DevNet, which uses `base58`, so we no longer need this check.


@joyqvq do you mind cleaning up the pre-0.18.0 code?

File: sdk/typescript/src/providers/json-rpc-provider.ts [hunk 9]
```diff
@@ -510,6 +534,9 @@ export class JsonRpcProvider extends Provider {
     digest: TransactionDigest
   ): Promise<SuiTransactionResponse> {
     try {
+      if (!isValidTransactionDigest(digest, "base64") && !isValidTransactionDigest(digest, "base58")) {
```
Comment (666lcz) [inline sdk/typescript/src/providers/json-rpc-provider.ts:537]:
```suggestion
      if (!isValidTransactionDigest(digest, "base58")) {
```

**Pull request merged**